docs/source/reference-io.rst

```diff
@@ -651,153 +651,110 @@ Spawning subprocesses
-
-        stdout_chunks = []
-        stderr_chunks = []
-
-        async with trio.Process(command, **options) as proc:
-
-            async def feed_input():
-                async with proc.stdin:
-                    if input:
-                        try:
-                            await proc.stdin.send_all(input)
-                        except trio.BrokenResourceError:
-                            pass
-
-            async def read_output(stream, chunks):
-                async with stream:
-                    while True:
-                        chunk = await stream.receive_some(32768)
-                        if not chunk:
-                            break
-                        chunks.append(chunk)
-
-            async with trio.open_nursery() as nursery:
-                if proc.stdin is not None:
-                    nursery.start_soon(feed_input)
-                if proc.stdout is not None:
-                    nursery.start_soon(read_output, proc.stdout, stdout_chunks)
-                if proc.stderr is not None:
-                    nursery.start_soon(read_output, proc.stderr, stderr_chunks)
-                await proc.wait()
-
-        stdout = b"".join(stdout_chunks) if proc.stdout is not None else None
-        stderr = b"".join(stderr_chunks) if proc.stderr is not None else None
-
-        if proc.returncode:
-            raise subprocess.CalledProcessError(
-                proc.returncode, proc.args, output=stdout, stderr=stderr
-            )
-        else:
-            return subprocess.CompletedProcess(
-                proc.args, proc.returncode, stdout, stderr
-            )
+The basic interface for running a subprocess start-to-finish is
+:func:`trio.run_process`.  It always waits for the subprocess to exit
+before returning, so there's no need to worry about leaving a process
+running by mistake after you've gone on to do other things.
+:func:`~trio.run_process` is similar to the standard library
+:func:`subprocess.run` function, but tries to have safer defaults:
+with no options, the subprocess's input is provided and its outputs
+are captured by the parent Trio process rather than connecting them
+to the user's terminal, and a failure in the subprocess will be propagated
+as a :exc:`subprocess.CalledProcessError` exception. Of course, these
+defaults can be changed where necessary.
+
+.. autofunction:: trio.run_process
+
+If you want to run a process with no standard stream redirections
+(i.e., taking input from the user and showing output and errors to the user),
+you can use ``trio.run_process(..., stdin=None, stdout=None, stderr=None)``.
+This runs the child in the same process group as the parent Trio process,
+so a Ctrl+C will be delivered to both the child process and the parent Trio
+process simultaneously. If that's not what you want, you'll need to delve
+into platform-specific APIs like (on UNIX) :func:`os.setpgrp` and
+:func:`os.tcsetpgrp`.
```

Comment:
1. Why is this up here, instead of in the `run_process` docstring? Output capturing semantics seems like a good thing to describe in the docstring :-).

2. I'm surprised at the switch to capturing by default. It doesn't seem obviously *wrong* to me, but it doesn't seem obviously right, either, and it's a pretty major change from `subprocess.run` so it needs some kind of rationale.

3. The mention of `setpgrp` and `tcsetpgrp` seems to come out of no-where here? I'm not sure if we should mention them at all, and if we do then probably it shouldn't be tacked onto the end of an unrelated paragraph :-)

Reply:
I implemented the `passthrough` option; let me know how you think it reads?

Reply:
The `passthrough` option is exactly the same as the stdlib's `capture_output`, but backwards, right? If we do do this, then making the default `capture_output=True` seems strictly better than making it `passthrough=False` :-).

I think I might still prefer `capture_stdout=True, capture_stderr=False`, but before we dive into that can you clarify the handling of `input`? In the stdlib `run`, AFAICT `input=None` (the default) means we let the child inherit the parent's stdin, and you have to pass `input=b""` if you want to give it an empty pipe. Are we matching that, or are you thinking that `passthrough` would affect this, or...?

Reply:
`passthrough=True` gives the stdlib defaults. `passthrough=False` is equivalent to `input=b""` if no `input` is specified, plus `capture_output=True`. I explained above in this thread why I think it's error-prone to have different isolation defaults for input vs output.

My preference would be to have one flag that controls "are the standard streams for this process basically isolated or are they basically inherited from the parent Trio process", and then people can modify the treatment of individual streams with the stdin/out/err options if they don't want all three streams treated the same way. If the standard streams are isolated and you didn't provide any input, you probably want the subprocess's input to come from /dev/null. It seems confusing to me to distinguish `input=None` from `input=b""` for that purpose.

Reply:
I've been thinking about this a lot this week!

So first, an anecdote: I actually ran into a real case where I would have used `run_process` if we had it. It was a very simple process monitor – I just wanted a script that would start up a flask dev server + a worker daemon that it talked to, and tear them both down if either exited. So:

```python
async def main():
    async with trio.open_nursery() as nursery:
        async def run_process_then_quit(cmd):
            await trio.run_process(cmd)
            nursery.cancel_scope.cancel()
        nursery.start_soon(run_process_then_quit, [...])
        nursery.start_soon(run_process_then_quit, [...])

trio.run(main)
```

Earlier you wrote:

> If you're writing a concurrent app, you might want to run multiple processes at once. That's maybe even _more_ likely than that you'd be careful not to run multiple processes at once. If you run multiple processes at once, it's not really safe to not isolate their I/O from each other.

This makes a lot of sense. But... OTOH, in my example, the correct behavior is actually to not capture stdout/stderr :-). Both daemons are dumping debug chatter on their output, and I definitely want to see all of it on my terminal as it happens. So I guess both capturing and non-capturing use cases do happen in real life, even when you're writing concurrent apps!

OTOH, in this case it probably would be better to do `input=b""`. It doesn't actually matter that much in practice, because these processes never read from their input. But if I was wrong about that, and they actually did, then `input=b""` would make sure I noticed and had to figure out some more reasonable solution. So that to me suggests that maybe we shouldn't be using a single flag to control both input and output handling simultaneously.

I've also seen *not* passing `input=b""` cause problems in real life, because of this kind of incorrect assumption. For example, this fun bug, which was caused because it never occurred to the pip developers that someone would write a `setup.py` that prompted the user for input: https://github.com/pypa/pip/issues/2732

One more random thought: I figured out why I'm a bit uncomfortable with either capturing both stdout + stderr or neither. It's because the the original motivation for making them different things was exactly to encourage people to capture stdout for programmatic consumption while leaving stderr pointing at a terminal :-). Specifically, to make unix pipelines work: `foo | bar` sends `foo`'s stdout to `bar`, but `foo`'s stderr stays on the terminal for a human to read.

OK. So how do we decide on defaults.

One good heuristic for defaults is: whatever configuration is most common, should be the default. I guess that's implicitly what we're thinking of when we debate whether concurrent programs normally want to capture or not.

There's another heuristic, to consider too, though: if there are two configurations for two sorts of users, then the default should be the one where the users two don't want it will realize immediately that they have the wrong configuration. I.e., any default will be wrong sometimes, so pick the default that when it's wrong, is obviously wrong.

In this case – the two kinds of users are the ones who want to capture output, and the ones who don't. The ones who want to capture output are going to immediately *do* something with that output – parse it or write it to a file or something. As they type `run_process`, they are thinking about how they're going to capture that output. If our default is to *not* capture output, then they will immediately notice and turn on output capturing. OTOH, the users who don't want to capture output may not be thinking about the output at all. They weren't planning to write any code to handle it. Maybe they expect that the program won't print any output at all. If our default *is* to capture output, and they don't stop and think about that, then they'll end up silently discarding the output. That seems like a nastier failure mode. Especially for the stderr case, where there really is no output for the first month your system is in production, and you don't notice that you're discarding the stderr until after things blow up and you have no idea why.

So I am leaning towards: we should follow `subprocess.run`, and *not* capture stdout/stderr by default.

If we go with this, then we'll need some spelling for how to override the default. For this I'm still leaning towards `capture_stdout=`, `capture_stderr=` as separate switches, because of the intuition from `foo | bar`.

Now, what about the input. Here, the argument works a bit differently... `input=b""` actually seems like the safer default. Having multiple processes reading from the same pipe at the same time is a nightmare; ttys have [some protection](http://curiousthing.org/sigttin-sigttou-deep-dive-linux) but it's still a mess. And even if you're only running one process, you can have fun failure modes like the bug above complaining that "pip just freezes on some packages". So I'm leaning towards `input=b""` as the default.

If we go with this, the nI'm not quite sure how to get the ergonomics right for when people want to override it. In particular:

* what do you write when you want the child to inherit stdin? `input=None`? Not super intuitive, though I guess it's straightforward enough to document and it would get the job done.  `stdin=0`? That'd be pretty arcane.

* How do we resolve conflicts between `input` and `stdin`? I think we want to allow `run_process(cmd, stdin=<arcane subprocess stuff>)`. But if `input=b""` is the default, then there's no way to tell for sure that someone hasn't pass both `input=` and `stdin=`, which is awkward. I guess the best option is just to say, if `input == b""`, and `stdin` was given, then we ignore `input`? Not the most elegant thing, but maybe the alternatives are worse.

There are also other options, like making the default signature `input=None` and declaring that that *means* the same thing as `input=b""` (except possibly if `stdin=` is passed). But that seems like even more of a mess than the alternatives.

What do you think?

Reply:
Good points all!

The danger of running multiple processes with the same stdout/stderr is that you get their output unhelpfully intermingled. You know, `process oprocess two outputne output` type of thing. :-) But most things use line buffering for terminal output, and even when they don't, a human can usually reconstruct what happened. So, I can see how letting multiple processes write to the same stdout/stderr could be more useful-in-practice that it is unsafe-in-theory.

Meanwhile, running multiple processes with the same _stdin_ is almost always a terrible idea; if there's contention on reading from it, you get a pretty bad user experience, like only half your keystrokes seeming to work.

I agree with your point about it being easy to notice you failed to capture the output you wanted to capture (because those bytes you're about to inspect will be `None` instead) and hard to notice you accidentally captured the output you didn't want to capture. It's actually really obvious with that lens that my previous default was misguided, so thanks for laying it out like that!

How about:
* The defaults are `stdin=subprocess.DEVNULL, stdout=None, stderr=None`. These can be overridden by the user passing `stdin`, `stdout`, `stderr` arguments.
* `capture_stdout=True` means `stdout=subprocess.PIPE` and errors if a different `stdout` was specified by the user. Likewise for `capture_stderr=True` and `stderr=subprocess.PIPE`.
* For `subprocess.run()` compatibility, `capture_output=True` is equivalent to `capture_stdout=True, capture_stderr=True`.
* `input=<some bytes>` means `stdin=subprocess.PIPE` and tells the parent Trio process to write those bytes to the pipe, then close it. `passthrough_stdin=True` means `stdin=None`. It's an error to specify more than one of `input`, `passthrough_stdin`, and `stdin`.
* `input=None` is treated equivalently to "you didn't specify any input" and still means `stdin=subprocess.DEVNULL`. This is a discrepancy from `subprocess.run()`, but at least it fails fast. If we don't like that, we could have `input=None` be an error, and make the default be some private marker object instead of `None` in order to be able to distinguish.

Reply:
> The danger of running multiple processes with the same stdout/stderr is that you get their output unhelpfully intermingled. You know, `process oprocess two outputne output` type of thing. :-) But most things use line buffering for terminal output, and even when they don't, a human can usually reconstruct what happened. So, I can see how letting multiple processes write to the same stdout/stderr could be more useful-in-practice that it is unsafe-in-theory.

I admit that in the real situation, the next thing I tried after getting the basic thing working was to mess around with proxying the process's output so I could make them different colors :-).

> [calling conventions]

Hmm.

I don't think we should allow or expose `PIPE` as a thing we let users see or pass. The pipes are an internal implementation detail, that the user can't get at, so we should abstract them away from the user's interface.

I kind of like `input=b""` better than `stdin=subprocess.DEVNULL`. Obviously they're equivalent semantically, but when we show the signature in the docs, `input=b""` seems easier to understand. Anyone who understands `input` in general will be able to figure out what `input=b""` means. But if we do `stdin=subprocess.DEVNULL`, then people still have to learn `input`, and then also know about the subprocess `stdin` arg (which is very complicated!), *and then* know what this weird `DEVNULL` thing is.

This thing where there are 3 different modes stdin could be in, and then two of the modes have more configuration (and one of them is really another like 7 modes, and includes passthrough as a special case)...... wedging this into Python's argument passing model is extremely awkward.

I almost wonder if we should drop `input`, and make it so `stdin` can be a bytestring, or else any of the things the `subprocess` would accept (except `PIPE`):

```python
def run_process(cmd, *, check=True, stdin=b"", capture_stdout=False, capture_stderr=False, **kwargs):
    ...
```

If we're splitting up `capture_output` and using the words `stdout` and `stderr`, then replacing `input` with `stdin` is nicely consistent.

This is where I paused to think more, and then left this comment window sitting open for several days :-). Thanks for the ping. I'm just going to hit post and see what you think :-).

Reply:
Quick preface -- I'm hoping this comment is helpful, from someone who's working to get trio into an internal app we use at work (but nonetheless in prod there, which is pretty neat). But I know there's work been done here and there's a whole trope about people coming into open source projects with demands and just yuck I'm really not trying to do that!

That being said, one thing I've found *extremely* useful is to, instead of passing a `capture_stdout` etc, is to pass a `stdout_redirect`, something like this:

```python
async def subproc_run(command, *, stdin_redirect=None, stdout_redirect=None,
                      stderr_redirect=None, **options):
    '''Any of stdin_redirect, stdout_redirect, and stderr_redirect
    should be trio channels. We'll read (stdin) or write (stdout/stderr)
    to those channels as needed.
    '''
```

Originally I started with the ``subprocess.run()`` replica [in the docs](https://trio.readthedocs.io/en/latest/reference-io.html#running-a-process-and-waiting-for-it-to-finish), but one of the ways I'm using this is to call into terraform, and those commands can sometimes take on the order of a minute to run. So I'd be sitting there looking at my console for, like, a long time, wondering what was up, and then finally at the end it would dump the whole stdout into the terminal, by looking into the returned ``subprocess.CompletedProcess.stdout`` (or stderr), which was suboptimal.

So instead I retooled it a bit to accept trio channels instead. I kinda punted on the "what's an appropriate buffer size here" question, and this is still in-process code (also, there's some stuff I'm doing with an ``async with capture_output()`` helper object that's probably too cutesy-clever for something like trio)  -- but anyways, I put the implementation and a few example usages into [this gist](https://gist.github.com/Badg/d188f2c0c0bb684a1a08920bdac0d3eb).

Reply:
Thanks for the suggestion! I actually thought about that as well when I was doing the most recent revision, and decided that without a specific request it was probably going to be rejected as unnecessary complexity. I guess we have a specific request now. :-)

@njsmith, what do you think about this idea of making `run_process` accept channels as `stdin`, `stdout`, and/or `stderr` arguments? If we do that we might want to add back the nursery.start() support, since it becomes a lot more reasonable for someone to want to do stuff with the Process object while `run_process` is executing.

Another option: we could keep `run_process` as the black-box start-to-finish API like it is currently, and add back something like the `async with open_process(...) as proc:` from a previous diff of mine, with channels support and so on.

This doesn't all need to happen on this diff, and I suspect it could be prototyped fine in an external library.

Reply:
@badg @oremanj What does "channels support" mean? What do these channels do? (Do you mean channels or streams?)

If you want to run a subprocess while incrementally running some arbitrary function on the output, then what is it that makes `async with trio.Process(..., stdout=PIPE) as p: ...` unpleasant and this better?

Reply:
I didn't actually have a chance to look at @Badg's code yet, but my impression of the proposal was to support saying something like
```
write_stdout, read_stdout = trio.open_memory_channel[bytes](0)
# start some task that will consume data from read_stdout
await trio.run_process(..., stdout=write_stdout)
```
I guess the main utility is that you can use incremental processing for some streams and bulk processing for others? And the channel interface is nicer than the stream one for incremental processing -- `async for chunk in channel:` rather than
```
while True:
    chunk = await stream.receive_some(ARBITRARILY_CHOSEN_POWER_OF_TWO)
    if not chunk:
        break
    ...
```

Reply:
That `while` loop is ugly no matter where you see it, but nothing stops us from adding a `receive_loop` method to our streams that lets us do an `async for chunk in stream.receive_loop(POWER_OF_TWO)`.

The other choice would be to wait for Python 3.8 and use that infamous `:=` thingy. 🤪

Reply:
> my impression of the proposal was to support saying something like

@njsmith @oremanj yep, that's the core of it -- I found it really useful to use channels (in my case, memory channels) as the high-level subprocess communication API. In my case the wrapper method doesn't even expose the ``capture_output``, ``input``, etc keywords; if you pass in channels, it captures the output there; if you don't, it discards it.

My use case is more eager processing than incremental processing, but that's sometimes a bit of a fine distinction, and I suspect that incremental processing is more likely to be the important one.

Reply:
Thinking about this some more... I think actually that the mental distinction I'm having here is between pre-allocating consumers and post-allocating consumers, given the ``run_process`` high-level API.

I think at an intuitive level, I expect ``async def run_process`` to be atomic -- it's responsible for starting, managing, and closing the process; when it returns, the process is definitely finished, and before you call it, the process definitely hasn't started (plus all of the other usual trio semantics).

So then if you want to combine that high-level API with eager/incremental processing of output/input, then you're (by definition) forced to preallocate your stdin producers and stdout/err consumers using a nursery, which you then need to coordinate with the ``run_process`` API somehow so that they get hooked up correctly. That, ultimately, is the API problem my wrapper is trying to solve.

Also partly relevant to the discussion in #959, I think my choice for a channel over a stream, even though the stream is quite literally the thing returned by the trio ``Process``, is the result of two things (listed in order of importance):

1. at least according to the docs, there's nothing suggesting I can pass an existing stream object to the ``trio.Process`` constructor, which means I'd have no way of preallocating the stream
2. it seems much, much more natural (and easier!) to use channels for this, **because of the ``send_channel, receive_channel = trio.open_memory_channel()`` method**, which immediately allocates everything you need to do this coordination (and close it down easily)

Hopefully that explanation of my train of thought makes sense!

Reply:
@Badg 
> at least according to the docs, there's nothing suggesting I can pass an existing stream object to the trio.Process constructor, which means I'd have no way of preallocating the stream

That's correct. It's possible in some exotic cases, but not in general.

> Thinking about this some more... I think actually that the mental distinction I'm having here is between pre-allocating consumers and post-allocating consumers, given the run_process high-level API.

Why do you prefer pre-allocating consumers, though? It seems like this is just as easy?

```python
async with trio.Process(...) as p:
    async with trio.open_nursery() as nursery:
        ...
```

Or if not, why not? (I guess two possibilities that come to mind: that doesn't do `check` and it doesn't kill the process on error. I sort of wonder if `Process.__aexit__` should kill the process if there's an exception, by analogy with nurseries cancelling siblings on error.)

Reply:
@njsmith I don't necessarily prefer preallocating consumers. The goal with my wrapper was to provide a high-level API that does ``check``, handles ``await process.wait()``, kills the process on error/cancellation/etc, all within a single ``await run_process`` call. As far as I'm aware, the only possible way to combine those semantics with incremental processing (of either input or output) is to preallocate, so that's what the API ends up exposing. Does that make sense?

Reply:
That's true, if the requirement is that it be a single `await my_run_process(...)`, then you have to preallocate. But what I'm still not sure of is: why is it important that it be a single `await my_run_process(...)`, as compared to a single `async with my_run_process(...): ...`?

Reply:
I mean, at the end of the day it's just my own personal API preference, so it deserves a heaping mound of salt. But I think my preference for an atomic ``my_run_process(...)`` over an async context manager, is that the context manager introduces a second concurrency mechanism to trio -- it's behaving both like a subprocess spawner, and a nursery, and my personal preference would be to keep those two separate. It was also my experience that the context manager resulted in more boilerplate, but I think that's the weaker point so I'm not sure it's worth exploring very far.

Like I said though, I'm just one person, and I always reserve the right to be wrong haha. Just thought I'd add some thoughts into the mix!